### PR TITLE
Fixes #13601, Better error messages

### LINKED ISSUE #13601: Better error message when users mistakenly import from named vs. default export
At google, we've had a number of users trip over import syntax when working with libraries that have default exports.

Given this scenario:

`has_named_export.ts`
```
export var a = 1;
```

`has_default_export.ts`
```
var b = '';
export default b;
```

1) Users often do:
```
import {b} from './has_default_export';
```
which gives:
`Error: module has no exported member 'b'`

Request: the error could be more helpful, like "did you mean to import the default export, with no curly braces"
This is especially likely when the name of the default export matches the named import.
Novice users look at `has_default_export` and think that B should be importable, and may never have worked with default exports.

2) Also possible:
```
import a from './has_named_export';
```
`"has_named_export" has no default export`

This one is less frequent, but it could say "did you mean to import the named import 'a' by using curly braces"

### PR DESCRIPTION
**Pull request by knrt10**
Signed-off-by: knrt10 <tripathi.kautilya@gmail.com>

Fixes #13601

### CODE REVIEW

**Comment by knrt10** [comment]
@RyanCavanaugh would you please suggest how to fix this [line error](https://github.com/Microsoft/TypeScript/pull/27505/files#diff-c3ed224e4daa84352f7f1abcd23e8ccaR1942). Its giving `Object is possibly 'undefined'.` 
I am not able to get if `moduleSymbol.exports` has `get` method, then why its not able to solve the issue. I have no previous experience in the repository and would be grateful if you could help me. 
Thanks in advance.

**Comment by knrt10** [comment]
Can someone please guide me how to update the tests for it.

**Comment by DanielRosenwasser** [comment]
```
jake runtests-parallel
jake baseline-accept
```

[See here](
https://github.com/Microsoft/TypeScript/blob/4ed85b7c44028970c7c3d3061ec2b1ab2bea3321/CONTRIBUTING.md#running-the-tests) for more details.

**Comment by knrt10** [comment]
@RyanCavanaugh would you please review

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -1946,6 +1949,13 @@ namespace ts {
                                     createDiagnosticForNode(suggestion.valueDeclaration, Diagnostics._0_is_declared_here, suggestionName)
                                 );
                             }
+
+                            if (defaultSymbolMatch) {
+                              error(name, Diagnostics.Module_0_has_no_exported_member_1_Did_you_mean_to_use_import_1_from_0_instead, getFullyQualifiedName(moduleSymbol), declarationNameToString(name));
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: 4 spaces per indentation please

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -1946,6 +1949,13 @@ namespace ts {
                                     createDiagnosticForNode(suggestion.valueDeclaration, Diagnostics._0_is_declared_here, suggestionName)
                                 );
                             }
+
+                            if (defaultSymbolMatch) {
+                              error(name, Diagnostics.Module_0_has_no_exported_member_1_Did_you_mean_to_use_import_1_from_0_instead, getFullyQualifiedName(moduleSymbol), declarationNameToString(name));
+                            }
+                            else {
+                              error(name, Diagnostics.Module_0_has_no_default_export_Did_you_mean_to_use_import_1_from_0_instead, getFullyQualifiedName(moduleSymbol), declarationNameToString(name));
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: 4 spaces per indentation please

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -1938,6 +1938,9 @@ namespace ts {
                         const moduleName = getFullyQualifiedName(moduleSymbol);
                         const declarationName = declarationNameToString(name);
                         const suggestion = getSuggestedSymbolForNonexistentModule(name, targetSymbol);
+                        const defaultSymbolMatch = moduleName.includes("default") ?
+                          symbolToString(resolveSymbol(moduleSymbol, dontResolveAlias)) === declarationName :
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: 4 spaces per indentation please

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -1938,6 +1938,9 @@ namespace ts {
                         const moduleName = getFullyQualifiedName(moduleSymbol);
                         const declarationName = declarationNameToString(name);
                         const suggestion = getSuggestedSymbolForNonexistentModule(name, targetSymbol);
+                        const defaultSymbolMatch = moduleName.includes("default") ?
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: `includes` is an ES6 library feature, so please use one of the helpers in `core.ts`

File: src/compiler/diagnosticMessages.json [hunk 5]
```diff
@@ -3898,7 +3898,14 @@
         "category": "Message",
         "code": 6502
     },
-
+    "Module '{0}' has no default export. Did you mean to use 'import {{1}} from \"{0}\"'instead?": {
+      "category": "Message",
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: 4 spaces per indentation please

File: src/compiler/diagnosticMessages.json [hunk 6]
```diff
@@ -3898,7 +3898,14 @@
         "category": "Message",
         "code": 6502
     },
-
+    "Module '{0}' has no default export. Did you mean to use 'import {{1}} from \"{0}\"'instead?": {
+      "category": "Message",
+      "code": 6503
+    },
+    "Module '{0}' has no exported member '{1}'. Did you mean to use 'import {1} from \"{0}\"'instead?": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: Can you add a test for this message?

File: tests/baselines/reference/exportSpellingSuggestion.errors.txt [hunk 7]
```diff
@@ -1,14 +1,17 @@
...
 
 ==== tests/cases/conformance/es6/modules/a.ts (0 errors) ====
     export function assertNever(x: never, msg: string) {
         throw new Error("Unexpected " + msg);
     }
     
-==== tests/cases/conformance/es6/modules/b.ts (1 errors) ====
+==== tests/cases/conformance/es6/modules/b.ts (2 errors) ====
     import { assertNevar } from "./a";
              ~~~~~~~~~~~
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/exportSpellingSuggestion.errors.txt]: You're issuing 2 errors on the same module now, so we should avoid that.

**Comment by DanielRosenwasser** [comment]
You can update tests with `jake baseline-accept` after running the tests with `jake runtests-parallel`.

**Comment by joaniedavis** [comment]
@DanielRosenwasser @knrt10 I'd like to pull this branch so I can update it with the requested review but I'm not familiar with forked repositories in GitHub. Do either of you have advice on what steps I need to take to get access to this branch of the code in order to make the requested updates?

**Comment by RyanCavanaugh** [comment]
@joaniedavis you'll have to open a new PR originating in your fork. You can add `knrt10`'s fork as a remote to your own local repo, check out the `fixErrorMessage` branch, then push that into your own fork - from there you can send a new PR with additional commits. Someone on the gitter can probably help out with specifics if needed.

Closing due to inactivity from OP.

**Pull request closed without merging**